### PR TITLE
Feat/refactoring

### PR DESCRIPTION
**Pull request by domhoward14**

### CODE REVIEW

**Comment by domhoward14** [comment]
@anilsenay I made updates based on your comments from the other [PR](https://github.com/go-swagno/swagno/pull/21) I made in the original repo. I did it here so after you're done with your PR we can merge them here, resolve conflicts, and then merge into the original swagno repo for release v1.0.0. 

I made some big changes which i think will make the api more efficient and easier to test and manage by having less code. Take a look if you can and let me know your thoughts 😁 .

File: generate.go [hunk 1]
```diff
@@ -2,41 +2,31 @@ package swagno
...
-		return false
-	}
-
-	numFields := rv.NumField()
-	return numFields > 0
-}
-
 func appendResponses(sourceResponses map[string]endpoint.JsonResponse, additionalResponses []response.Info) map[string]endpoint.JsonResponse {
-	for _, response := range additionalResponses {
+	responseGenerator := generator.NewResponseGenerator()
```
Comment (anilsenay) [inline generate.go]: You commented about making this more idiomatic, can you do it in your PR?

**Comment by anilsenay** [comment]
LGTM! I will try to finish implementation of supporting interface issue which I mentioned yesterday. Its getting complicated in code 😄 If I can not finish it in today, I will open a pull request anyway to show what I have done so far. So I think we can merge your pull request for now but I want try to solve interface issue in today

File: generate.go (hunk 1 above)
Comment (domhoward14) [reply generate.go]: yup, i'll get that done today

**Comment by domhoward14** [comment]
> LGTM! I will try to finish implementation of supporting interface issue which I mentioned yesterday. Its getting complicated in code 😄 If I can not finish it in today, I will open a pull request anyway to show what I have done so far. So I think we can merge your pull request for now but I want try to solve interface issue in today

Awesome, sounds good! And yeah, if you can throw up what you're working on i can pull it down and try to test it on my end just so i can understand the problem you're solving a little better.

**Comment by anilsenay** [comment]
https://github.com/domhoward14/swagno/pull/3#discussion_r1414932061
https://github.com/domhoward14/swagno/pull/3#discussion_r1414940538
https://github.com/domhoward14/swagno/pull/3#discussion_r1414929631

The refactors based on your comments can be made in this PR so I think I can close #3

**Comment by anilsenay** [comment]
I also suggest add these parts from my recent pr into this pr:

- https://github.com/domhoward14/swagno/pull/5/files#diff-60d1e6c254f4ae0135dfdc6720fcdbdf7f35fac60770f5fc6dff13ebb8d00556R84
- https://github.com/domhoward14/swagno/pull/5/files#diff-b7b70ada8477e17dcaa6f14a85b7f284d0a24a39e6c7463a5fc5609266096070R92-R95 (if it make sense to you)
- https://github.com/domhoward14/swagno/pull/5/files#diff-b7b70ada8477e17dcaa6f14a85b7f284d0a24a39e6c7463a5fc5609266096070R145-R150 (prevent recursion based on https://github.com/go-swagno/swagno/issues/15)

**Comment by domhoward14** [comment]
@anilsenay When you can could you please take a look at the updates? I made some more idiomatic changes to function and type names after reading [this](https://google.github.io/styleguide/go/best-practices.html),. There were some things in there that I totally forgot about and was the reference for a lot of these changes.  Plus I made the other changes we discussed from the other PRs. Let me know what you think when you get a chance 😁 .

**Comment by anilsenay** [comment]
@domhoward14 hey, namings are fine to me. Nicely done

But you removed `CustomResponseType` I guess. I think we should add it back or add something similar. Like I said before, adding two methods(`Description` and `ReturnCode`) for each struct is not something I want in my projects.

---

In addition, did you check [these](https://github.com/domhoward14/swagno/pull/4#issuecomment-1846707130)? What do you think?
> I also suggest add these parts from my recent pr into this pr:
> 
>     * https://github.com/domhoward14/swagno/pull/5/files#diff-60d1e6c254f4ae0135dfdc6720fcdbdf7f35fac60770f5fc6dff13ebb8d00556R84
> 
>     * https://github.com/domhoward14/swagno/pull/5/files#diff-b7b70ada8477e17dcaa6f14a85b7f284d0a24a39e6c7463a5fc5609266096070R92-R95 (if it make sense to you)
> 
>     * https://github.com/domhoward14/swagno/pull/5/files#diff-b7b70ada8477e17dcaa6f14a85b7f284d0a24a39e6c7463a5fc5609266096070R145-R150 (prevent recursion based on [Detect Recursive structs go-swagno/swagno#15](https://github.com/go-swagno/swagno/issues/15))

**Comment by domhoward14** [comment]
> @domhoward14 hey, namings are fine to me. Nicely done
> 
> But you removed `CustomResponseType` I guess. I think we should add it back or add something similar. Like I said before, adding two methods(`Description` and `ReturnCode`) for each struct is not something I want in my projects.


Ah, my bad. I may have misunderstood. I thought we said that we were going to try and implement all the custom-type support in a future version. After I looked into the `CustomResponseType` (shown below) and the code to supporting  it added a lot of untested complexity due to it having an `any` type for the `Model attribute. Plus using only one struct type with different arguments to model different responses API looked complicated since the reflection sees it all as one type, which led to your solution modifying with hashes, which solves the distinction problem but introduces less user client control over their type name, and hurts the readability when the definition is rendered on the swagger UI. 

So my question is do you have a solution where we can keep this custom struct type which creates all the definitions, while keeping the name distinct and readable, and without introducing support for a blank interface type (i.e. `any`)? I have some ideas bouncing off of what you did for `CustomResponseType`, but I just think it's going to take some more time and I don't want to quickly push a release that could lead to a lot of complexity while we're trying to get a release out.

```
type CustomResponseType struct {
	Model       any
	ReturnCode  string
	Description string
}
```

**Comment by domhoward14** [comment]
> 
> In addition, did you check [these](https://github.com/domhoward14/swagno/pull/4#issuecomment-1846707130)? What do you think?
> 
> > I also suggest add these parts from my recent pr into this pr:
> > ```
> > * https://github.com/domhoward14/swagno/pull/5/files#diff-60d1e6c254f4ae0135dfdc6720fcdbdf7f35fac60770f5fc6dff13ebb8d00556R84
> > 
> > * https://github.com/domhoward14/swagno/pull/5/files#diff-b7b70ada8477e17dcaa6f14a85b7f284d0a24a39e6c7463a5fc5609266096070R92-R95 (if it make sense to you)
> > 
> > * https://github.com/domhoward14/swagno/pull/5/files#diff-b7b70ada8477e17dcaa6f14a85b7f284d0a24a39e6c7463a5fc5609266096070R145-R150 (prevent recursion based on [Detect Recursive structs go-swagno/swagno#15](https://github.com/go-swagno/swagno/issues/15))
> > ```

And thanks for the reminder for the first and third link. I'll get that added in tonight. For the second one though I left a comment regarding it [here](https://github.com/domhoward14/swagno/pull/5/files#r1419990366).

**Comment by domhoward14** [comment]
Pushed up another update addressing comments @anilsenay

**Comment by anilsenay** [comment]
> > @domhoward14 hey, namings are fine to me. Nicely done
> > But you removed `CustomResponseType` I guess. I think we should add it back or add something similar. Like I said before, adding two methods(`Description` and `ReturnCode`) for each struct is not something I want in my projects.
> 
> Ah, my bad. I may have misunderstood. I thought we said that we were going to try and implement all the custom-type support in a future version. After I looked into the `CustomResponseType` (shown below) and the code to supporting it added a lot of untested complexity due to it having an `any` type for the `Model attribute. Plus using only one struct type with different arguments to model different responses API looked complicated since the reflection sees it all as one type, which led to your solution modifying with hashes, which solves the distinction problem but introduces less user client control over their type name, and hurts the readability when the definition is rendered on the swagger UI.
> 
> So my question is do you have a solution where we can keep this custom struct type which creates all the definitions, while keeping the name distinct and readable, and without introducing support for a blank interface type (i.e. `any`)? I have some ideas bouncing off of what you did for `CustomResponseType`, but I just think it's going to take some more time and I don't want to quickly push a release that could lead to a lot of complexity while we're trying to get a release out.
> 
> ```
> type CustomResponseType struct {
> 	Model       any
> 	ReturnCode  string
> 	Description string
> }
> ```

Actually, it won't add more complexity. The hashing solution is for another issue, not related to this. We will add only a few lines of code for supporting `CustomResponseType`. It will be basicly like that if the input model is type of `CustomResponseType` then use only `input.Model` by calling same function with it. 
For example:
```go
func (g DefinitionGenerator) CreateDefinition(t interface{}) {
...
case reflect.Struct:
		if reflectReturn == reflect.TypeOf(response.CustomResponseType{}) {
			// if CustomResponseType, use Model struct in it
			g.CreateDefinition(t.(response.CustomResponseType).Model)
			return
		}
		properties = g.createStructDefinitions(reflectReturn, t)
...
```

In summary: if the input type of `CustomResponseType` then use CustomResponseType.Model, else use input itself

Isn't it the same way as current state?

**Comment by domhoward14** [comment]
> > > @domhoward14 hey, namings are fine to me. Nicely done
> > > But you removed `CustomResponseType` I guess. I think we should add it back or add something similar. Like I said before, adding two methods(`Description` and `ReturnCode`) for each struct is not something I want in my projects.
> > 
> > 
> > Ah, my bad. I may have misunderstood. I thought we said that we were going to try and implement all the custom-type support in a future version. After I looked into the `CustomResponseType` (shown below) and the code to supporting it added a lot of untested complexity due to it having an `any` type for the `Model attribute. Plus using only one struct type with different arguments to model different responses API looked complicated since the reflection sees it all as one type, which led to your solution modifying with hashes, which solves the distinction problem but introduces less user client control over their type name, and hurts the readability when the definition is rendered on the swagger UI.
> > So my question is do you have a solution where we can keep this custom struct type which creates all the definitions, while keeping the name distinct and readable, and without introducing support for a blank interface type (i.e. `any`)? I have some ideas bouncing off of what you did for `CustomResponseType`, but I just think it's going to take some more time and I don't want to quickly push a release that could lead to a lot of complexity while we're trying to get a release out.
> > ```
> > type CustomResponseType struct {
> > 	Model       any
> > 	ReturnCode  string
> > 	Description string
> > }
> > ```
> 
> Actually, it won't add more complexity. The hashing solution is for another issue, not related to this. We will add only a few lines of code for supporting `CustomResponseType`. It will be basicly like that if the input model is type of `CustomResponseType` then use only `input.Model` by calling same function with it. For example:
> 
> ```go
> func (g DefinitionGenerator) CreateDefinition(t interface{}) {
> ...
> case reflect.Struct:
> 		if reflectReturn == reflect.TypeOf(response.CustomResponseType{}) {
> 			// if CustomResponseType, use Model struct in it
> 			g.CreateDefinition(t.(response.CustomResponseType).Model)
> 			return
> 		}
> 		properties = g.createStructDefinitions(reflectReturn, t)
> ...
> ```
> 
> In summary: if the input type of `CustomResponseType` then use CustomResponseType.Model, else use input itself
> 
> Isn't it the same way as current state?

Ah ok, thanks for explaining that. I didn't realize that the custom type support and `CustomResponseType` were independent of each other. That solution in your last response looks good. I'll put that back in today and push back up for your review @anilsenay .

**Comment by anilsenay** [comment]
@domhoward14 alright 🔥 Also can you add some tests for `CustomResponseType` in tests you wrote?

Yep, they are totally dfferent topics. Supporting maps, interfaces etc. won't be included in this release, as we discussed. We would find a better way later.

Let me help you to find the code should be added for `CustomResponseType`
https://github.com/anilsenay/swagno-v1/blob/refactor-definition-2/http/response/response.go#L9-L28
https://github.com/anilsenay/swagno-v1/blob/refactor-definition-2/generator/definition.go#L60-L64
https://github.com/anilsenay/swagno-v1/blob/refactor-definition-2/generate.go#L21-L26

**Comment by domhoward14** [comment]
> @domhoward14 alright 🔥 Also can you add some tests for `CustomResponseType` in tests you wrote?
> 
> Yep, they are totally dfferent topics. Supporting maps, interfaces etc. won't be included in this release, as we discussed. We would find a better way later.
> 
> Let me help you to find the code should be added for `CustomResponseType` https://github.com/anilsenay/swagno-v1/blob/refactor-definition-2/http/response/response.go#L9-L28 https://github.com/anilsenay/swagno-v1/blob/refactor-definition-2/generator/definition.go#L60-L64 https://github.com/anilsenay/swagno-v1/blob/refactor-definition-2/generate.go#L21-L26

@anilsenay Awesome, thanks for the links and feedback! Made all the updates plus I refactored to use the way you designed as the primary way. After I understood it better and was updating the README I couldn't think of a situation where someone would probably want to create their own `Description()` and `ReturnCode()` functions, so in the read me the official recommended way is to use `response.New(models.EmptySuccessfulResponse{}, "OK", "200")` instead of making your structs have to implement the interface. So thanks for this addition. I added it to the testing and I think it's working out and looking great!

If you have any questions or concerns about the updates lmk 😁 !

**Comment by anilsenay** [comment]
@domhoward14 LGTM 🚀 Thanks for all your efford really 🔥 

So I think its ready to merge now. Then we can merge the PR in the original repo and prepare a release.

**Review by anilsenay — APPROVED** [review]

**Comment by domhoward14** [comment]
> @domhoward14 LGTM 🚀 Thanks for all your efford really 🔥
> 
> So I think its ready to merge now. Then we can merge the PR in the original repo and prepare a release.

Aweshum!! I know this took a little longer than we initially talked about (way more than a couple of days lol), but I really think it was worth it and i'm pretty happy with how clients interact with the API to make swagger docs. Good job on the back and forth! I'll get it merged now so we can get this release out Leggggooo 🚀 !

**Comment by anilsenay** [comment]
@domhoward14 ahaha ye I didnt expect this to take longer 😄

**Pull request merged**